### ISSUE TITLE
Bug: [no-unnecessary-boolean-literal-compare] False positive and unsafe fix with exactOptionalPropertyTypes

### ISSUE THREAD

**Issue by andersk** [typescript-eslint/typescript-eslint#9105]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.2&fileType=.tsx&code=MYewdgzgLgBCAOUCW4IC4YG8YDMA2AhgOYD8GARiCHgKYFgwC%2BMAZFroaRVbfTAD4wArmAAmNHEjA1RTGAF4sjANwAoUJGo0AdHhBEAFAmSpt%2BYgvmKcBPBBoBKNUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ARhw5IhTWo3yJoQ%2BLTIcAtsRGJ0UJdA7RI4MAF8QZoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEk6AHgIbR64HGOoAKmsiWeAJ4AVQfzKU8mAK7oANOBqQSU5CwpgpsxQF8QOoA&tokens=false

### Repro Code

```TypeScript
const options: { flag?: boolean } & { flag?: boolean | undefined } = {};
console.log(options.flag === false);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "exactOptionalPropertyTypes": true,
    "strict": true
  }
}
```

### Expected Result

No errors. I haven’t changed the default options `{ allowComparingNullableBooleansToTrue: true, allowComparingNullableBooleansToFalse: true }`, so comparing `boolean | undefined` to `false` should be allowed.

### Actual Result

```
  2:13  error  This expression unnecessarily compares a boolean value to a boolean instead of using it directly  @typescript-eslint/no-unnecessary-boolean-literal-compare
```

The rule automatically applies an incorrect fix `options.flag === false` → `!options.flag`, which changes the output of the code from `false` to `true`.

**Comment by bradzacher** [comment]
This is actually working as expected.

When you intersect two union types the resulting type is the intersection of the unions - i.e. the types that exist in both unions.
For example - `type T = (number | undefined) & number;` - `T === number`.

When you intersect two object types TS will include all the unique properties between the two types, and then it will intersect the common properties. So what you end up with in your example is `boolean & (boolean | undefined)` - i.e. `boolean`.

Which is why you can write [`options.flag.valueOf()` without any guards and TS will not error](https://www.typescriptlang.org/play/?exactOptionalPropertyTypes=true#code/CYUwxgNghgTiAEYD2A7AzgF3kgDhglqmgFzwDe8AZtAOYD8pARkkhCFCvAL7wBk5VWg3jNW7TgB94AVxShK+FCGDcA3AFgAULgJEAdNSg09ANygRpIAPKUAFAEoNmoA).

[When we inspect the types for `.flag`, TS tells us its type is just `boolean`](https://www.typescriptlang.org/play/?exactOptionalPropertyTypes=true#code/CYUwxgNghgTiAEYD2A7AzgF3kgDhglqmgFzwDe8AZtAOYD8pARkkhCFCvAL7wBk5VWg3jNW7TgB94AVxShK+FCGDcA3AFgAULgJEAdNSg0NmgPSn4lq-AB6dIA). It says there's no possibility of it being `undefined`. Hence the rule errors.

[If you tweak the types slightly you'll see that rules like `no-unnecessary-condition` will also error here](https://typescript-eslint.io/play/#ts=5.3.2&fileType=.tsx&code=MYewdgzgLgBCAOUCW4IC4YG8YDMA2AhgOYD8GOBeEApjAL4wBkWuhp5lNMAPjAK5gAJtRxIw1QfRgBeLHQDcAKCQ4YACgTJUAOnzEZ02RSrUAlCzqKgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ARhw5IhTWo3yJoQ%2BLTIcAtsRGJ0UJdA7RIAGnDY8RUsgrU6DZm07de-QSLHqmAE0r5KHEx6kAZGkOYAviARQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEk6AHgIbR64HGOoAKmsiWeAJ4AVQfzKU8mAK7oANOBqQSU5CwpgpsxQF8QOoA&tokens=false) because the types tell us there's no `undefined`.

IDK if the behaviour is correct from TS's side here - surely it should be erroring on the property / saying the type is `undefined` at read time?
I don't know for sure - would require an issue to discuss with them.

**Comment by andersk** [comment]
Both members of the intersection have `flag?: …`, not `flag: …`. So the property is allowed to be absent.

> Which is why you can write [options.flag.valueOf() without any guards and TS will not error](https://www.typescriptlang.org/play/?exactOptionalPropertyTypes=true#code/CYUwxgNghgTiAEYD2A7AzgF3kgDhglqmgFzwDe8AZtAOYD8pARkkhCFCvAL7wBk5VWg3jNW7TgB94AVxShK+FCGDcA3AFgAULgJEAdNSg09ANygRpIAPKUAFAEoNmoA).

This does seem like a TypeScript bug.

> [When we inspect the types for .flag, TS tells us its type is just boolean](https://www.typescriptlang.org/play/?exactOptionalPropertyTypes=true#code/CYUwxgNghgTiAEYD2A7AzgF3kgDhglqmgFzwDe8AZtAOYD8pARkkhCFCvAL7wBk5VWg3jNW7TgB94AVxShK+FCGDcA3AFgAULgJEAdNSg0NmgPSn4lq-AB6dIA). It says there's no possibility of it being undefined. Hence the rule errors.

That shows `(property) flag?: boolean`. The `?` indicates that TypeScript knows the property might be absent.

**Comment by kirkwaiblinger** [comment]
@bradzacher but granted it's an optional property the two intersection parts are basically identical unless you have the `--exactOptionalPropertyTypes` flag right?

FWIW our code just doesn't handle intersections at all, see
https://github.com/typescript-eslint/typescript-eslint/blob/584db29ec44ce4e9cb71afac35d48994889168e6/packages/eslint-plugin/src/rules/no-unnecessary-boolean-literal-compare.ts#L123-L126

Also, side note, this seems like the type of case that would benefit from assignability checking a la https://github.com/microsoft/TypeScript/pull/56448. Is there an issue on our side for introducing that into the codebase now that it's landing/landed?

**Comment by andersk** [comment]
It looks like this is the related TypeScript bug:

- microsoft/TypeScript#58174

**Comment by kirkwaiblinger** [comment]
Disregard my comment, didn't realize `--exactOptionalPropertyTypes` _is_ set

**Comment by bradzacher** [comment]
> That shows (property) flag?: boolean

Yes - that's because TS's "intellisense printer" shows you the type AND the context. So when it shows you the context (the property name) it also needs to show the "divider". When it prints the divider it also inspects the flags for the property to determine if it should also print the `?` in addition to the standard `:`.

---

microsoft/TypeScript#58174 looks to be the culprit - that's why it works without the intersection but breaks it with it.

So I'll close this as a bug in TS.
